### ISSUE TITLE
useQuerySelector

### ISSUE THREAD

**Issue by shtse8** [vueuse/vueuse#1985]
### Clear and concise description of the problem

Sometimes, we want to select an element from document but it may not exist or will changes in some time.
We need to select the element in reactive.

### Suggested solution

```ts
function useQuerySelector(selectors: string): Ref<HTMLElement | undefined> {
  const value = ref<HTMLElement | undefined>(document.head?.querySelector<HTMLElement>(selectors) ?? undefined)
  useMutationObserver(document.documentElement, (mutations) => {
    const latestMutation = mutations[0]
    if (latestMutation.type === 'childList')
      value.value = document?.querySelector<HTMLElement>(selectors) ?? undefined
  }, {
    childList: true,
    subtree: true,
  })
  return value
}
```

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vueuse/vueuse/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vueuse.org/guide).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by chaii3** [comment]
Is it possible to use template refs for this goal ?

**Comment by shtse8** [comment]
> template refs

no probably, in my use case, I am querying element outside component like head tags 
 value for sharing.

**Comment by chaii3** [comment]
I'm really don't know your case but maybe useTitle might help you?

I'm sorry for this questions.

In my opinion change element from other component is magic. That's why I think this composition isn't a vue way. High chance I'm wrong it's only my opinion.

Maybe we need to make composition like useMeta.

**Comment by shtse8** [comment]
> I'm really don't know your case but maybe useTitle might help you?
> 
> I'm sorry for this questions.
> 
> In my opinion change element from other component is magic. That's why I think this composition isn't a vue way. High chance I'm wrong it's only my opinion.
> 
> Maybe we need to make composition like useMeta.

agree, if we can watch the head directly. I find no way to get the latest meta in reactive manner instead of selecting the element to watch.
https://github.com/vueuse/head/issues/76

**Comment by chaii3** [comment]
We can incapsulate selecting meta elements in composition if we don't find better solution.

**Comment by shtse8** [comment]
> We can incapsulate selecting meta elements in composition if we don't find better solution.

Here is my workaround solution to watch the meta. some metas are not like title which is always exist. 

```ts
function useHead(tag: string, props: Record<string, string> = {}, attributeName?: string): Ref<string | undefined> {
  if (isClient) {
    const element = useQuerySelector(`${tag}${Object.entries(props).map(x => `[${x[0]}="${x[1]}"]`).join('')}`)
    return useElementValue(element, attributeName)
  }
  else {
    return getHead(tag, props, x => attributeName ? x[attributeName] : x.children)
  }
}
export function useElementValue(el: MaybeElementRef, attributeName?: string): Ref<string | undefined> {
  const value = ref<string | undefined>((attributeName ? unrefElement(el)?.getAttribute(attributeName) : unrefElement(el)?.textContent) ?? undefined)
  useMutationObserver(el, (mutations) => {
    const latestMutation = mutations[0]
    switch (latestMutation.type) {
      case 'attributes':
        if (attributeName && latestMutation.attributeName === attributeName)
          value.value = unrefElement(el)?.getAttribute(attributeName) ?? undefined
        break
      case 'childList':
        value.value = unrefElement(el)?.textContent ?? undefined
        break
    }
  }, attributeName
    ? {
        attributeFilter: [attributeName],
        attributes: true,
      }
    : {
        childList: true,
        subtree: true,
      })
  return value
}

export function useTitle(): Ref<string | undefined> {
  return useHead('title')
}

export function useMeta(props: { name?: string; property?: string }): Ref<string | undefined> {
  return useHead('meta', props, 'content')
}

export function useDescription(): Ref<string | undefined> {
  return useMeta({ name: 'description' })
}

export function useOg(name: string): Ref<string | undefined> {
  return useMeta({ property: `og:${name}` })
}
```


I am making a sharer component to share the current page with title, description and url. There is `useTitle` for watching title. but not others. Also, I'd like to add some fallbacks since `og:title`, which is aim for sharing and may not contain site name, may be different from `title`. So if there is `og:title`, share if, if not, then title.


```

**Comment by chaii3** [comment]
@shtse8 is https://github.com/vueuse/head can resolve this problem ?

**Comment by shtse8** [comment]
> @shtse8 is https://github.com/vueuse/head can resolve this problem ?

no, that what I am using. I need to watch the head. but the head is not watchable.
https://github.com/vueuse/head/issues/76

**Comment by ngocsangyem** [comment]
Thank @shtse8, this composable is very helpful because I also face the same issue. Do you have any advance for `useQuerySelector`?

**Comment by shtse8** [comment]
> Thank @shtse8, this composable is very helpful because I also face the same issue. Do you have any advance for `useQuerySelector`?

Actually, you typically don't need to query an element unless it is controlled by a third-party script. If the element is managed by Vue, you should always stay within the reactive paradigm. In Vue, the DOM is simply a result of the virtual DOM (vnode) changes, and manipulating the DOM directly can lead to inconsistencies and unexpected behavior.

When you keep your interactions within the reactive system, Vue can efficiently manage updates and keep your application state synchronized with the UI. Directly querying and manipulating the DOM should be avoided unless you have a very specific reason and fully understand the implications. It can bypass Vue's reactivity system and potentially cause performance issues or bugs.

In summary, always prefer using Vue's reactive tools and components to manage the state and UI, and only query the DOM directly if absolutely necessary and with a clear understanding of the consequences.

**Comment by KazimirPodolski** [comment]
This feature would be very useful as the case when a piece of DOM is NOT managed by the user is the very situation when we could use a clean helper.

DOM within Vue is not a counter-argument for this feature as it's just not the target of the feature.
